plugin/cache/handler.go

```
@@ -38,9 +38,13 @@ func (c *Cache) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
 	if i != nil {
 		ttl = i.ttl(now)
 	}
-	if i == nil {
+	if i == nil || (ttl < 0 && c.verifyStale) {
+		// try to resolve upstream if we have no cached record or the cached one is stale and serve_stale is set to verify.
 		crr := &ResponseWriter{ResponseWriter: w, Cache: c, state: state, server: server, do: do}
-		return c.doRefresh(ctx, state, crr)
+		rcode, err := c.doRefresh(ctx, state, crr)
+		if i == nil || (rcode == dns.RcodeSuccess || rcode == dns.RcodeNameError){
+			return rcode, err
+		}
```

Calling `c.doRefresh` generally results in sending a response to the client if a response is received - no matter what the response code is. As proposed, if the response code is not NXDOMAIN or NOERROR, the client can be sent two responses, the first one being the current upstream response, and the second one (which would be ignored by the client) being the stale cached entry.

---

OK, I realise that I did not understand your comment in https://github.com/coredns/coredns/pull/4756#discussion_r672453981.

I wrapped the response writer as you suggested, so that it only writes if the response is nxdomain or noerror. I'm not quite sure that the behaviour is correct in case the underlying `ResponseWriter` returns an error to `WriteMsg`. As proposed the code just ignores the error and proceeds with serving stale entry.